experimental/traffic-portal/src/app/api/topology.service.ts

```diff
@@ -0,0 +1,196 @@
+		return this.post<ResponseTopology>("topologies", topology).toPromise();
+	}
+
+	/**
+	 * Replaces an existing Topology with the provided new definition of a
+	 * Topology.
+	 *
+	 * @param topology The full new definition of the Topology being updated
+	 */
+	public async updateTopology(topology: ResponseTopology): Promise<ResponseTopology> {
+		return this.put<ResponseTopology>(`topologies?name=${topology.name}`, topology).toPromise();
+	}
+
+	/**
+	 * Generates a material tree from a topology.
+	 *
+	 * @param topology The topology to generate a material tree from.
+	 * @returns a material tree.
+	 */
+	public topologyToTree(topology: ResponseTopology): Array<TopTreeNode> {
+		const treeNodes: Array<TopTreeNode> = [];
+		const topLevel: Array<TopTreeNode> = [];
+		for (const node of topology.nodes) {
+			const name = node.cachegroup;
+			const cachegroup = node.cachegroup;
+			const children: Array<TopTreeNode> = [];
+			const parents: Array<TopTreeNode> = [];
+			treeNodes.push({
+				cachegroup,
+				children,
+				name,
+				parents,
+			});
+		}
+		for (let index = 0; index < topology.nodes.length; index++) {
+			const node = topology.nodes[index];
+			const treeNode = treeNodes[index];
+			if (!(node.parents instanceof Array) || node.parents.length < 1) {
+				topLevel.push(treeNode);
+				continue;
+			}
+			for (const parent of node.parents) {
+				treeNodes[parent].children.push(treeNode);
+				treeNode.parents.push(treeNodes[parent]);
+			}
+		}
+		return topLevel;
+	}
+
+	/**
+	 * Generates a topology from a material tree.
+	 *
+	 * @param name The topology name
+	 * @param description The topology description
+	 * @param treeNodes The data for a material tree
+	 * @returns a material tree.
+	 */
+	public treeToTopology(name: string, description: string, treeNodes: Array<TopTreeNode>): ResponseTopology {
+		const topologyNodeIndicesByCacheGroup: Map<string, number> = new Map();
+		const nodes: Array<ResponseTopologyNode> = new Array<ResponseTopologyNode>();
+		this.treeToTopologyInner(topologyNodeIndicesByCacheGroup, nodes, undefined, treeNodes);
+		const topology: ResponseTopology = {
+			description,
+			lastUpdated: new Date(),
+			name,
+			nodes,
+		};
+		return topology;
+	}
+
+	/**
+	 * Inner recursive function for generating a Topology from a material tree.
+	 *
+	 * @param topologyNodeIndicesByCacheGroup A map of Topology node indices
+	 * using cache group names as the key
+	 * @param topologyNodes The mutable array of Topology nodes
+	 * @param parent The parent, if it exists
+	 * @param treeNodes The data for a material tree
+	 */
+	protected treeToTopologyInner(topologyNodeIndicesByCacheGroup: Map<string, number>,
+		topologyNodes: Array<ResponseTopologyNode>, parent: ResponseTopologyNode | undefined, treeNodes: Array<TopTreeNode>): void {
+
+		for (const treeNode of treeNodes) {
+			const cachegroup = treeNode.cachegroup;
+			const parents: number[] = [];
+			if (parent instanceof Object) {
```

Comment:
You should pretty much never use `instanceof Object` because that doesn't tell you much of anything:
```javascript
[] instanceof Object // true
new Number() instanceof Object // true
new Date() instanceof Object // true
(() => {}) instanceof Object // true
```
If you want to know if something is an `object` primitive type, use `typeof(x) === "object"`, if you want to know if it's an instance of some specific class, you can use `x instanceof Class`, but nearly everything is an `Object`.

Note also some weird behavior with null values:
```javascript
typeof(null) // "object"
null instanceof Object // false
```

Reply:
> If you want to know if something is an object primitive type, use typeof(x) === "object"

`typeof object` is bad for the reason you mentioned, that `typeof null` returns `"object"`.

> if you want to know if it's an instance of some specific class, you can use `x instanceof Class`

How? If I change that line to

```typescript
			if (parent instanceof ResponseTopologyNode) {
```

I get

```
TS2693:  ResponseTopologyNode  only refers to a type, but is being used as a value here.
```

But since all I'm trying to rule out is non-instances, `parent instanceof Object` seems like the safest way.

Reply:
I would invert the check. In general, `typeof` is less prone to weird JavaScript-isms than `instanceof`. So instead, I'd write
```javascript
if (typeof(parent) !== "undefined")
```

Reply:
> I would invert the check. In general, `typeof` is less prone to weird JavaScript-isms than `instanceof`. So instead, I'd write
> 
> ```js
> if (typeof(parent) !== "undefined")
> ```

Using `typeof` in caa96123e0, but without the parenthesis, because `typeof` is an operator, not a function